File: src/main/java/com/now/nowbot/service/MessageServiceImpl/QualifiedMapService.java
package com.now.nowbot.service.MessageServiceImpl;

import com.now.nowbot.NowbotApplication;
import com.now.nowbot.model.JsonData.Search;
import com.now.nowbot.model.enums.OsuMode;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.ImageService;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.OsuApiService.OsuBeatmapApiService;
import com.now.nowbot.throwable.ServiceException.QualifiedMapException;
import com.now.nowbot.util.Pattern4ServiceImpl;
import jakarta.annotation.Resource;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.regex.Matcher;

@Service("QUALIFIEDMAP")
public class QualifiedMapService implements MessageService<Matcher> {
    @Resource
    OsuBeatmapApiService beatmapApiService;
    @Resource
    ImageService         imageService;

    @Override
    public boolean isHandle(MessageEvent event, DataValue<Matcher> data) {
        var m = Pattern4ServiceImpl.QUALIFIEDMAP.matcher(event.getRawMessage().trim());
        if (m.find()) {
            data.setValue(m);
            return true;
        } else return false;
    }
    @Override
    public void HandleMessage(MessageEvent event, Matcher matcher) throws Throwable {
        // 获取参数
        var mode_str = matcher.group("mode");
        var status = matcher.group("status");
        var sort = matcher.group("sort");
        var range_str = matcher.group("range");
        short range;
        short mode = OsuMode.DEFAULT.getModeValue();

        if (mode_str != null) mode = OsuMode.getMode(mode_str).getModeValue();
        if (status == null) status = "q";
        if (sort == null) sort = "ranked_asc";

        if (range_str == null) {
            range = 12; //从0开始
        } else {
            try {
                range = Short.parseShort(range_str);
            } catch (Exception e) {
                throw new QualifiedMapException(QualifiedMapException.Type.Q_Parameter_Error);
            }
        }

        if (range < 1 || range > 999) throw new QualifiedMapException(QualifiedMapException.Type.Q_Parameter_OutOfRange);

        int page = 1;
        int page_aim = (int) Math.max(Math.floor(range / 50f) + 1, 10);// 这里需要重复获取，page最多取10页（500个），总之我不知道怎么实现

        var query = new HashMap<String, Object>();
        status = getStatus(status);
        query.put("m", mode);
        query.put("s", status);
        query.put("sort", getSort(sort));
        query.put("page", page);

        try {
            Search data = null;
            int resultCount = 0;
            do {
                if (data == null) {
                    data = beatmapApiService.searchBeatmap(query);
                    resultCount += data.getBeatmapsets().size();
                    continue;
                }
                page ++;
                query.put("page", page);
                var result = beatmapApiService.searchBeatmap(query);
                resultCount += result.getBeatmapsets().size();
                data.getBeatmapsets().addAll(result.getBeatmapsets());
            } while (resultCount < data.getTotal() && page < page_aim);

            if (resultCount == 0) throw new QualifiedMapException(QualifiedMapException.Type.Q_Result_NotFound);

            data.setResultCount(Math.min(data.getTotal(), range));
            data.setRule(status);
            data.sortBeatmapDiff();
            var img = imageService.getPanelA2(data);
            event.getSubject().sendImage(img);
        } catch (Exception e) {
            NowbotApplication.log.error("QuaMap: ", e);
            throw new QualifiedMapException(QualifiedMapException.Type.Q_Send_Error);
        }
    }

    private static String getStatus(String status) {
        return switch (status.toLowerCase()) {
            case "0", "p" -> "pending";
            case "1", "r" -> "ranked";
            case "2", "a" -> "approved";
            case "4", "l" -> "loved";
            case "-1", "5", "w" -> "wip";
            case "-2", "6", "g" -> "graveyard";
            default -> "qualified";
        };
    }

    private static String getSort(String sort) {
        return switch (sort.toLowerCase()) {
            case "t", "t+", "ta", "title", "title asc", "title_asc" -> "title_asc";
            case "t-", "td", "title desc", "title_desc" -> "title_desc";
            case "a", "a+", "aa", "artist", "artist asc", "artist_asc" -> "artist_asc";
            case "a-", "ad", "artist desc", "artist_desc" -> "artist_desc";
            case "d", "d+", "da", "difficulty", "difficulty asc", "difficulty_asc", "s", "s+", "sa", "star", "star asc", "star_asc" ->
                    "difficulty_asc";
            case "d-", "dd", "difficulty desc", "difficulty_desc", "s-", "sd", "star desc", "star_desc" ->
                    "difficulty_desc";
            case "m", "m+", "ma", "map", "rating", "rating asc", "rating_asc" -> "rating_asc";
            case "m-", "md", "map desc", "rating desc", "rating_desc" -> "rating_desc";
            case "p", "p+", "pa", "plays", "pc asc", "plays asc", "plays_asc" -> "plays_asc";
            case "p-", "pd", "pc desc", "plays desc", "plays_desc" -> "plays_desc";
            case "r", "r+", "ra", "ranked", "time asc", "ranked asc", "ranked_asc" -> "ranked_asc";
            case "r-", "rd", "time desc", "ranked desc", "ranked_desc" -> "ranked_desc";
            default -> "relevance_desc";
        };
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/QualifiedMapService.kt
package com.now.nowbot.service.messageServiceImpl

import com.now.nowbot.dao.BindDao
import com.now.nowbot.entity.ServiceCallStatistic
import com.now.nowbot.model.enums.CoverType
import com.now.nowbot.model.enums.OsuMode
import com.now.nowbot.model.osu.BeatmapsetSearch
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.ImageService
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.osuApiService.OsuBeatmapApiService
import com.now.nowbot.service.osuApiService.OsuUserApiService

import com.now.nowbot.throwable.botRuntimeException.IllegalArgumentException
import com.now.nowbot.throwable.botRuntimeException.IllegalStateException
import com.now.nowbot.util.AsyncMethodExecutor
import com.now.nowbot.util.DataUtil
import com.now.nowbot.util.Instruction
import com.now.nowbot.util.command.FLAG_MODE
import com.now.nowbot.util.command.FLAG_PAGE
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.regex.Matcher

@Service("QUALIFIED_MAP") class QualifiedMapService(
    private val beatmapApiService: OsuBeatmapApiService,
    private val userApiService: OsuUserApiService,
    private val imageService: ImageService,
    private val bindDao: BindDao
) : MessageService<QualifiedMapService.QualifiedMapParam> {

    data class QualifiedMapParam(
        val mode: OsuMode,
        val status: String?,
        val genre: Byte?,
        val sort: String,
        val page: Int = 1,
    )

    override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<QualifiedMapParam>,
    ): Boolean {
        val matcher = Instruction.QUALIFIED_MAP.matcher(messageText)

        if (!matcher.find()) {
            return false
        }

        data.value = getParam(event, matcher)
        return true
    }

    override fun handleMessage(event: MessageEvent, param: QualifiedMapParam): ServiceCallStatistic? {

        val query = mapOf<String, Any>(
            "m" to param.mode.modeValue,
            "s" to (param.status?.takeIf { it != "any" } ?: "qualified"),
            "sort" to param.sort,
            "page" to 1,
        )

        val search: BeatmapsetSearch

        try {
            search = beatmapApiService.parallelSearchBeatmapset(query)

            val maxPerPage: Int = if (search.resultCount <= 24) {
                12
            } else {
                48
            }

            val split = DataUtil.splitPage(search.beatmapsets, param.page, maxPerPage)

            // 后处理
            AsyncMethodExecutor.awaitTripleCallableExecute(
                { beatmapApiService.applyBeatmapsetRankedTime(split.first) },
                { userApiService.applyUserForBeatmapset(split.first) },
                {
                    if (maxPerPage == 12) {
                        // 给完整面板整点头像
                        userApiService.asyncDownloadAvatarFromBeatmapsets(split.first)
                    } else {
                        beatmapApiService.asyncDownloadCoverFromSets(split.first, CoverType.LIST_2X)
                    }
                }
            )

            search.beatmapsets = split.first

            val img = imageService.getPanel(mapOf(
                "search" to search,
                "page" to split.second,
                "max_page" to split.third
            ), "A2")
            event.reply(img)
        } catch (e: Exception) {
            log.error("过审谱面：", e)
            throw IllegalStateException.Send("过审谱面")
        }

        return ServiceCallStatistic.builds(event,
            beatmapsetIDs = search.beatmapsets.map { it.beatmapsetID },
        )
    }

    private fun getParam(event: MessageEvent, matcher: Matcher): QualifiedMapParam {
        // 获取参数
        val statusStr: String? = matcher.group("status")
        val sortStr: String? = matcher.group("sort")
        val genreStr: String? = matcher.group("genre")
        val pageStr: String? = matcher.group(FLAG_PAGE)

        val page = pageStr?.toIntOrNull() ?: 1

        if (page !in 1..999) {
            throw IllegalArgumentException.WrongException.Henan()
        }

        val mode = OsuMode.getMode(OsuMode.getMode(matcher.group(FLAG_MODE)), bindDao.getGroupModeConfig(event))
        val status = DataUtil.getStatus(statusStr)
        val genre = DataUtil.getGenre(genreStr)
        val sort = DataUtil.getSort(sortStr ?: "ranked_asc")

        return QualifiedMapParam(
            mode = mode,
            status = status,
            genre = genre,
            sort = sort,
            page = page
        )
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(QualifiedMapService::class.java)
    }
}
